bug: 디벨롭이 안켜집니다

## Bug

### **버그를 설명해주세요**

- src/achievements/achievements.service.ts:36:49 - error TS2339: Property 'getUserTrash' does not exist on type 'CacheService'.
- src/cache/cache.service.ts:86:14 - error TS2304: Cannot find name 'TrashSummary'.
- src/cache/cache.service.ts:86:14 - error TS4055: Return type of public method from exported class has or is using private name 'TrashSummary'.

### **맞는 동작**

서버 켜지기

--- a/package.json
+++ b/package.json
@@ -34,6 +34,7 @@
     "class-validator": "^0.13.2",
     "config": "^3.3.7",
     "cors": "^2.8.5",
+    "date-fns": "^2.29.2",
     "passport": "^0.6.0",
     "passport-jwt": "^4.0.0",
     "passport-kakao": "^1.0.1",

--- a/src/achievements/achievement.interceptor.ts
+++ b/src/achievements/achievement.interceptor.ts
@@ -1,0 +1,31 @@
+import {
+  CallHandler,
+  ExecutionContext,
+  Inject,
+  Injectable,
+} from '@nestjs/common';
+import { map } from 'rxjs';
+import { CacheService } from 'src/cache/cache.service';
+
+@Injectable()
+export class AchievementInterceptor {
+  constructor(
+    @Inject(CacheService) private readonly cacheService: CacheService,
+  ) {}
+
+  intercept(context: ExecutionContext, next: CallHandler) {
+    if (context.getType() === 'http') {
+      const userId = context.switchToHttp().getRequest().user?.id;
+      return next.handle().pipe(
+        map(async (data) => {
+          return {
+            data,
+            achievement: userId
+              ? await this.cacheService.getAchievementNotifications(userId)
+              : null,
+          };
+        }),
+      );
+    }
+  }
+}

--- a/src/achievements/achievements.controller.ts
+++ b/src/achievements/achievements.controller.ts
@@ -1,0 +1,19 @@
+import { Controller, Get, Param, ParseIntPipe } from '@nestjs/common';
+import { AchievementsRepository } from './achievements.repository';
+
+@Controller('achievements')
+export class AchievementsController {
+  constructor(
+    private readonly achievementsRepository: AchievementsRepository,
+  ) {}
+
+  @Get()
+  async getAchievements() {
+    return this.achievementsRepository.getAchievements();
+  }
+
+  @Get(':id')
+  async getAchievement(@Param('id', ParseIntPipe) id) {
+    return this.achievementsRepository.getAchievement(id);
+  }
+}

--- a/src/achievements/achievements.module.ts
+++ b/src/achievements/achievements.module.ts
@@ -1,0 +1,19 @@
+import { Module, forwardRef } from '@nestjs/common';
+import { AchievementsService } from './achievements.service';
+import { AchievementsController } from './achievements.controller';
+import { CacheModule } from 'src/cache/cache.module';
+import { DatabaseModule } from 'src/database/database.module';
+import { AchievementsRepository } from './achievements.repository';
+
+@Module({
+  imports: [
+    forwardRef(() => {
+      return CacheModule;
+    }),
+    DatabaseModule,
+  ],
+  controllers: [AchievementsController],
+  providers: [AchievementsService, AchievementsRepository],
+  exports: [AchievementsRepository],
+})
+export class AchievementsModule {}

--- a/src/achievements/achievements.repository.ts
+++ b/src/achievements/achievements.repository.ts
@@ -1,0 +1,153 @@
+import { Injectable } from '@nestjs/common';
+import { database } from 'src/common/environments';
+import { DatabaseService } from 'src/database/database.service';
+import { Achievement } from './entity/achievement.entity';
+import { UserAchievement } from './entity/user-achievement.entity';
+
+@Injectable()
+export class AchievementsRepository {
+  constructor(private readonly databaseService: DatabaseService) {}
+
+  async getAchievement(id: number): Promise<Achievement> {
+    const query = `
+      SELECT
+        "id",
+        "name",
+        "description",
+        "imageUri"
+      FROM
+        ${database.tables.achievement}
+      WHERE
+        id = $1
+    `;
+
+    const result = await this.databaseService.query<{
+      id: string;
+      name: string;
+      description: string;
+      imageUri: string;
+    }>(query, [id]);
+
+    if (result.length === 0) {
+      return null;
+    } else {
+      const { id, ...ret } = result[0];
+      return {
+        id: parseInt(id),
+        ...ret,
+      };
+    }
+  }
+
+  async getAchievements(): Promise<Achievement[]> {
+    const query = `
+      SELECT
+        "id",
+        "name",
+        "description",
+        "imageUri"
+      FROM
+        ${database.tables.achievement}
+    `;
+    const result = await this.databaseService.query<{
+      id: string;
+      name: string;
+      description: string;
+      imageUri: string;
+    }>(query);
+    return result.map((achievement) => {
+      const { id, ...ret } = achievement;
+      return {
+        id: parseInt(id),
+        ...ret,
+      };
+    });
+  }
+
+  async getUserAchievements(userId: number): Promise<UserAchievement[]> {
+    const query = `
+      SELECT
+        "id",
+        "name",
+        "description",
+        "imageUri",
+        "achievedAt"
+      FROM
+        ${database.tables.achievement}
+        LEFT JOIN
+        ${database.tables.achiever}
+        ON
+        ${database.tables.achievement}."id" = ${database.tables.achiever}."achievementId"
+        AND
+        ${database.tables.achiever}."userId" = $1
+    `;
+
+    const result = await this.databaseService.query<{
+      id: string;
+      name: string;
+      description: string;
+      imageUri: string;
+      achievedAt: Date;
+    }>(query, [userId]);
+
+    return result.map((achievement) => {
+      const { id, ...ret } = achievement;
+      return {
+        id: parseInt(id),
+        ...ret,
+      };
+    });
+  }
+
+  async getAchievableIds(userId: number): Promise<number[]> {
+    const query = `
+      SELECT
+        id
+      FROM
+        ${database.tables.achievement}
+        LEFT JOIN
+        ${database.tables.achiever}
+        ON
+        ${database.tables.achievement}."id" = ${database.tables.achiever}."achievementId"
+      WHERE
+        "userId" = $1
+        AND
+        "achievedAt" IS NULL
+    `;
+    const result = await this.databaseService.query<{ id: string }>(query, [
+      userId,
+    ]);
+    return result.map((row) => {
+      return parseInt(row.id);
+    });
+  }
+
+  async isAchieved(userId: number, achievementId: number): Promise<boolean> {
+    const query = `
+      SELECT
+        "achievedAt"
+      FROM
+        ${database.tables.achiever}
+      WHERE
+        "userId" = $1
+        AND
+        "achievementId" = $2
+    `;
+    const result = await this.databaseService.query<{ achievedAt: Date }>(
+      query,
+      [userId, achievementId],
+    );
+    return result.length === 1;
+  }
+
+  async unlock(userId: number, achievementId: number): Promise<void> {
+    const query = `
+      INSERT INTO
+        ${database.tables.achiever}
+        ("userId", "achievementId", "achievedAt")
+      VALUES
+        ($1, $2, NOW())
+    `;
+    await this.databaseService.query(query, [userId, achievementId]);
+  }
+}

--- a/src/achievements/achievements.service.ts
+++ b/src/achievements/achievements.service.ts
@@ -1,0 +1,67 @@
+import { Injectable } from '@nestjs/common';
+import { CacheService } from 'src/cache/cache.service';
+import { AchievementsRepository } from './achievements.repository';
+
+@Injectable()
+export class AchievementsService {
+  private readonly plasticAchievements = [
+    { id: 1, criterion: 100 },
+    { id: 2, criterion: 1000 },
+    { id: 3, criterion: 10000 },
+  ];
+  private readonly canAchievements = [
+    { id: 4, criterion: 100 },
+    { id: 5, criterion: 1000 },
+    { id: 6, criterion: 10000 },
+  ];
+  private readonly paperAchievements = [
+    { id: 7, criterion: 100 },
+    { id: 8, criterion: 1000 },
+    { id: 9, criterion: 10000 },
+  ];
+
+  private readonly petAchievements = [
+    { id: 10, criterion: 100 },
+    { id: 11, criterion: 1000 },
+    { id: 12, criterion: 10000 },
+  ];
+
+  constructor(
+    private readonly cacheService: CacheService,
+    private readonly achievementsRepository: AchievementsRepository,
+  ) {}
+
+  async unlockAchievement(id: number, userId: number) {
+    await Promise.all([
+      this.achievementsRepository.unlock(userId, id),
+      this.cacheService.updateAchievementNotification(userId, id),
+    ]);
+  }
+
+  async checkTrashAchievements(userId: number) {
+    const trashCounts = await this.cacheService.getUserTrashAllSummary(userId);
+    const ids = await this.achievementsRepository.getAchievableIds(userId);
+    await Promise.all([
+      ...this.plasticAchievements.map(({ id, criterion }) => {
+        if (ids.includes(id) && trashCounts.plastic.success >= criterion) {
+          return this.unlockAchievement(id, userId);
+        }
+      }),
+      ...this.canAchievements.map(({ id, criterion }) => {
+        if (ids.includes(id) && trashCounts.can.success >= criterion) {
+          return this.unlockAchievement(id, userId);
+        }
+      }),
+      ...this.paperAchievements.map(({ id, criterion }) => {
+        if (ids.includes(id) && trashCounts.paper.success >= criterion) {
+          return this.unlockAchievement(id, userId);
+        }
+      }),
+      ...this.petAchievements.map(({ id, criterion }) => {
+        if (ids.includes(id) && trashCounts.pet.success >= criterion) {
+          return this.unlockAchievement(id, userId);
+        }
+      }),
+    ]);
+  }
+}

--- a/src/achievements/entity/achievement.entity.ts
+++ b/src/achievements/entity/achievement.entity.ts
@@ -3,5 +3,4 @@ export class Achievement {
   name: string;
   description: string;
   imageUri: string;
-  achievedAt: Date;
 }

--- a/src/achievements/entity/user-achievement.entity.ts
+++ b/src/achievements/entity/user-achievement.entity.ts
@@ -1,0 +1,6 @@
+import { PartialType } from '@nestjs/swagger';
+import { Achievement } from './achievement.entity';
+
+export class UserAchievement extends PartialType(Achievement) {
+  achievedAt: Date;
+}

--- a/src/app.module.ts
+++ b/src/app.module.ts
@@ -6,6 +6,9 @@ import { DatabaseModule } from './database/database.module';
 import { CacheModule } from './cache/cache.module';
 import { TrashcansModule } from './trashcans/trashcans.module';
 import { TrashModule } from './trash/trash.module';
+import { AchievementsModule } from './achievements/achievements.module';
+import { APP_INTERCEPTOR } from '@nestjs/core';
+import { AchievementInterceptor } from './achievements/achievement.interceptor';
 
 @Module({
   imports: [
@@ -13,12 +16,19 @@ import { TrashModule } from './trash/trash.module';
     CacheModule,
     DatabaseModule,
     UsersModule,
+    AchievementsModule,
     ConfigModule.forRoot({
       isGlobal: true,
       envFilePath: '.development.env',
     }),
     TrashcansModule,
     TrashModule,
+  ],
+  providers: [
+    {
+      provide: APP_INTERCEPTOR,
+      useClass: AchievementInterceptor,
+    },
   ],
 })
 export class AppModule {}

--- a/src/cache/cache.module.ts
+++ b/src/cache/cache.module.ts
@@ -26,8 +26,9 @@ export class CacheModule implements OnApplicationShutdown {
 
   async onApplicationBootstrap() {
     await this.moduleRef.get('REDIS_CLIENT').connect();
-    this.cacheService.migrateUsersPoint();
-    this.cacheService.migrateUsersTrash();
+    await this.cacheService.migrateUsersPoint();
+    await this.cacheService.migrateUsersAllTrash();
+    await this.cacheService.migrateUsersUsageTrialTrash();
   }
 
   onApplicationShutdown() {

--- a/src/cache/cache.module.ts
+++ b/src/cache/cache.module.ts
@@ -26,6 +26,7 @@ export class CacheModule implements OnApplicationShutdown {
 
   async onApplicationBootstrap() {
     await this.moduleRef.get('REDIS_CLIENT').connect();
+    await this.cacheService.flushAll();
     await this.cacheService.migrateUsersPoint();
     await this.cacheService.migrateUsersAllTrash();
     await this.cacheService.migrateUsersUsageTrialTrash();

--- a/src/cache/cache.module.ts
+++ b/src/cache/cache.module.ts
@@ -1,4 +1,5 @@
-import { Module, OnApplicationShutdown } from '@nestjs/common';
+import { AchievementsModule } from 'src/achievements/achievements.module';
+import { Module, OnApplicationShutdown, forwardRef } from '@nestjs/common';
 import { ModuleRef } from '@nestjs/core';
 import { createClient } from 'redis';
 import { redis } from 'src/common/environments';
@@ -6,7 +7,12 @@ import { DatabaseModule } from 'src/database/database.module';
 import { CacheService } from './cache.service';
 
 @Module({
-  imports: [DatabaseModule],
+  imports: [
+    DatabaseModule,
+    forwardRef(() => {
+      return AchievementsModule;
+    }),
+  ],
   providers: [
     CacheService,
     {

--- a/src/cache/cache.service.ts
+++ b/src/cache/cache.service.ts
@@ -172,12 +172,14 @@ export class CacheService {
       }),
     );
 
-    const userTrashLogsInOneUsage = this.client.hKeys(
+    const userTrashLogsInOneUsage = await this.client.hGetAll(
       `user-trash:${userId}-${open}`,
     );
+
     const type = Object.keys(userTrashLogsInOneUsage)[0]?.split('-')[0];
     const success = +(userTrashLogsInOneUsage[`${type}-success`] ?? 0);
     const failure = +(userTrashLogsInOneUsage[`${type}-failure`] ?? 0);
+
     return {
       Date: open,
       type,

--- a/src/cache/cache.service.ts
+++ b/src/cache/cache.service.ts
@@ -1,17 +1,19 @@
 import { Inject, Injectable } from '@nestjs/common';
 import { RedisClientType as Client } from 'redis';
-import { database } from 'src/common/environments';
+import { database, redis } from 'src/common/environments';
 import { DatabaseService } from 'src/database/database.service';
 import { TrashSummary } from 'src/trash/dto/trash-summary.dto';
 import { format } from 'date-fns';
 import { ko } from 'date-fns/locale';
 import { OneTrialTrashSummary } from 'src/trash/dto/one-trial-trash-summary.dto';
+import { AchievementsRepository } from 'src/achievements/achievements.repository';
 
 @Injectable()
 export class CacheService {
   constructor(
     @Inject('REDIS_CLIENT') private client: Client,
     private readonly databaseService: DatabaseService,
+    private readonly achievementsRepository: AchievementsRepository,
   ) {}
 
   async flushAll() {
@@ -249,5 +251,24 @@ export class CacheService {
 
   async updateUserPoint(userId: number, change: number) {
     await this.client.zIncrBy('user-point', change, userId.toString());
+  }
+
+  async updateAchievementNotification(userId: number, achievementId: number) {
+    await this.client.sAdd(
+      `${redis.keys.achievementNotification}:${userId}`,
+      achievementId.toString(),
+    );
+  }
+
+  async getAchievementNotifications(userId: number) {
+    const achievementIds = await this.client.sMembers(
+      `${redis.keys.achievementNotification}:${userId}`,
+    );
+    await this.client.del(`${redis.keys.achievementNotification}:${userId}`);
+    return Promise.all(
+      achievementIds.map((id) => {
+        return this.achievementsRepository.getAchievement(+id);
+      }),
+    );
   }
 }

--- a/src/cache/cache.service.ts
+++ b/src/cache/cache.service.ts
@@ -5,6 +5,7 @@ import { DatabaseService } from 'src/database/database.service';
 import { TrashSummary } from 'src/trash/dto/trash-summary.dto';
 import { format } from 'date-fns';
 import { ko } from 'date-fns/locale';
+import { OneTrialTrashSummary } from 'src/trash/dto/one-trial-trash-summary.dto';
 
 @Injectable()
 export class CacheService {
@@ -59,38 +60,29 @@ export class CacheService {
 
   async migrateUsersUsageTrialTrash() {
     const usersTrashUsage = await this.databaseService.query<{
-      userId: string;
+      userId: number;
       trashcanId: string;
       open: Date;
       close: Date;
     }>(
       `SELECT "userId", "trashcanId", "open", "close" FROM ${database.tables.trashcanUsage};`,
     );
-
-    try {
-      await Promise.all(
-        usersTrashUsage.map(({ userId, open }) => {
-          const openString = format(open, 'yyyy-MM-dd HH:mm:ss.SSS', {
-            locale: ko,
-          });
-          const key = `user-trash:${userId}-${open.getFullYear()}-${
-            open.getMonth() + 1
-          }`;
-          return this.client.rPush(key, openString);
-        }),
-      );
-    } catch (error) {
-      console.log(error);
-    }
-
+    await Promise.all(
+      usersTrashUsage.map(({ userId, open }) => {
+        const openString = format(open, 'yyyy-MM-dd HH:mm:ss.SSS', {
+          locale: ko,
+        });
+        const key = `user-trash:${userId}-${open.getFullYear()}-${
+          open.getMonth() + 1
+        }`;
+        return this.client.rPush(key, openString);
+      }),
+    );
     const usersTrashLogsInTrial = (
       await Promise.all(
         usersTrashUsage.map(({ userId, open, close }) => {
-          const openString = format(open, 'yyyy-MM-dd HH:mm:ss.SSS', {
-            locale: ko,
-          });
-          const closeString = format(close, 'yyyy-MM-dd HH:mm:ss.SSS', {
-            locale: ko,
+          const [openString, closeString] = [open, close].map((date) => {
+            return format(date, 'yyyy-MM-dd HH:mm:ss.SSS', { locale: ko });
           });
           return this.databaseService.query<{
             userId: string;
@@ -99,8 +91,8 @@ export class CacheService {
             ok: string;
           }>(
             `SELECT "userId", "type", "at", "ok" FROM ${database.tables.trash} 
-              WHERE "userId" = ${userId} 
-              AND "at" BETWEEN '${openString}' AND '${closeString}';`,
+                WHERE "userId" = ${userId} 
+                AND "at" BETWEEN '${openString}' AND '${closeString}';`,
           );
         }),
       )
@@ -131,14 +123,14 @@ export class CacheService {
     userId: number,
     year: number,
     month: number,
-  ): Promise<any> {
+  ): Promise<OneTrialTrashSummary[]> {
     const userTrashUsagesInMonth = await this.client.lRange(
       `user-trash:${userId}-${year}-${month}`,
       0,
       -1,
     );
 
-    const userTrashLogsInMonth = (
+    return (
       await Promise.all(
         userTrashUsagesInMonth.map((open) => {
           return this.client.hGetAll(`user-trash:${userId}-${open}`);
@@ -149,19 +141,21 @@ export class CacheService {
       const success = +(userTrashLogsInOneUsage[`${type}-success`] ?? 0);
       const failure = +(userTrashLogsInOneUsage[`${type}-failure`] ?? 0);
       return {
-        Date: new Date(userTrashUsagesInMonth[index]),
+        date: new Date(userTrashUsagesInMonth[index]),
         type,
         success,
         failure,
       };
     });
-    return userTrashLogsInMonth;
   }
 
-  async addUserUsageTrialTrash(userId, open, close) {
-    const openString = format(open, 'yyyy-MM-dd HH:mm:ss.SSS', { locale: ko });
-    const closeString = format(close, 'yyyy-MM-dd HH:mm:ss.SSS', {
-      locale: ko,
+  async addUserUsageTrialTrash(
+    userId: number,
+    open: Date,
+    close: Date,
+  ): Promise<OneTrialTrashSummary> {
+    const [openString, closeString] = [open, close].map((date) => {
+      return format(date, 'yyyy-MM-dd HH:mm:ss.SSS', { locale: ko });
     });
     const trashLogs = await this.databaseService.query<{
       userId: string;
@@ -173,24 +167,20 @@ export class CacheService {
           WHERE "userId" = ${userId} 
           AND "at" BETWEEN '${openString}' AND '${closeString}';`,
     );
+    let type: string,
+      success = 0,
+      failure = 0;
     await Promise.all(
-      trashLogs.map(({ type, ok }) => {
+      trashLogs.map(({ type: trashType, ok }) => {
         const key = `user-trash:${userId}-${open}`;
-        const field = `${type}-${ok ? 'success' : 'failure'}`;
+        const field = `${trashType}-${ok ? 'success' : 'failure'}`;
+        type = trashType;
+        ok ? success++ : failure++;
         return this.client.hIncrBy(key, field, 1);
       }),
     );
-
-    const userTrashLogsInOneUsage = await this.client.hGetAll(
-      `user-trash:${userId}-${open}`,
-    );
-
-    const type = Object.keys(userTrashLogsInOneUsage)[0]?.split('-')[0];
-    const success = +(userTrashLogsInOneUsage[`${type}-success`] ?? 0);
-    const failure = +(userTrashLogsInOneUsage[`${type}-failure`] ?? 0);
-
     return {
-      Date: open,
+      date: open,
       type,
       success,
       failure,
@@ -225,7 +215,7 @@ export class CacheService {
     });
   }
 
-  async getUserTrashSummary(userId: number): Promise<TrashSummary> {
+  async getUserTrashAllSummary(userId: number): Promise<TrashSummary> {
     const userTrashSummary = await this.client.hGetAll(`user-trash:${userId}`);
     return {
       can: {
@@ -247,14 +237,13 @@ export class CacheService {
     };
   }
 
-  async updateUserPoint(userId: number, change: number) {
-    await this.client.zIncrBy('user-point', change, userId.toString());
+  async updateUserTrashAllSummary(userId, type, successNum, failureNum) {
+    const key = `user-trash:${userId}`;
+    await this.client.hIncrBy(key, `${type}-success`, successNum);
+    await this.client.hIncrBy(key, `${type}-failure`, failureNum);
   }
 
-  updateUserTrash(userId: number, trashType: string, isCorrect: boolean) {
-    const today = new Date();
-    const key = `user-trash:${userId}-${today.getUTCFullYear()}-${today.getUTCMonth()}`;
-    const field = `${trashType}-${isCorrect ? 'success' : 'failure'}`;
-    this.client.hIncrBy(key, field, 1);
+  async updateUserPoint(userId: number, change: number) {
+    await this.client.zIncrBy('user-point', change, userId.toString());
   }
 }

--- a/src/cache/cache.service.ts
+++ b/src/cache/cache.service.ts
@@ -3,6 +3,8 @@ import { RedisClientType as Client } from 'redis';
 import { database } from 'src/common/environments';
 import { DatabaseService } from 'src/database/database.service';
 import { TrashSummary } from 'src/trash/dto/trash-summary.dto';
+import { format } from 'date-fns';
+import { ko } from 'date-fns/locale';
 
 @Injectable()
 export class CacheService {
@@ -33,7 +35,7 @@ export class CacheService {
     }
   }
 
-  async migrateUsersTrash() {
+  async migrateUsersAllTrash() {
     const usersTrash = await this.databaseService.query<{
       userId: string;
       type: string;
@@ -43,8 +45,128 @@ export class CacheService {
       SELECT "userId", "type", "at", "ok" FROM ${database.tables.trash};
     `);
     return Promise.all(
-      usersTrash.map(({ userId, type, at, ok }) => {
-        const key = `user-trash:${userId}-${at.getUTCFullYear()}-${at.getUTCMonth()}`;
+      usersTrash.map(({ userId, type, ok }) => {
+        const key = `user-trash:${userId}`;
+        const field = `${type}-${ok ? 'success' : 'failure'}`;
+        return this.client.hIncrBy(key, field, 1);
+      }),
+    );
+  }
+
+  async migrateUsersUsageTrialTrash() {
+    const usersTrashUsage = await this.databaseService.query<{
+      userId: string;
+      trashcanId: string;
+      open: Date;
+      close: Date;
+    }>(
+      `SELECT "userId", "trashcanId", "open", "close" FROM ${database.tables.trashcanUsage};`,
+    );
+
+    try {
+      await Promise.all(
+        usersTrashUsage.map(({ userId, open }) => {
+          const openString = format(open, 'yyyy-MM-dd HH:mm:ss.SSS', {
+            locale: ko,
+          });
+          const key = `user-trash:${userId}-${open.getFullYear()}-${open.getMonth()}`;
+          return this.client.rPush(key, openString);
+        }),
+      );
+    } catch (error) {
+      console.log(error);
+    }
+
+    const usersTrashLogsInTrial = (
+      await Promise.all(
+        usersTrashUsage.map(({ userId, open, close }) => {
+          const openString = format(open, 'yyyy-MM-dd HH:mm:ss.SSS', {
+            locale: ko,
+          });
+          const closeString = format(close, 'yyyy-MM-dd HH:mm:ss.SSS', {
+            locale: ko,
+          });
+          return this.databaseService.query<{
+            userId: string;
+            type: string;
+            at: Date;
+            ok: string;
+          }>(
+            `SELECT "userId", "type", "at", "ok" FROM ${database.tables.trash} 
+              WHERE "userId" = ${userId} 
+              AND "at" BETWEEN '${openString}' AND '${closeString}';`,
+          );
+        }),
+      )
+    ).map((trashLogs, index) => {
+      return {
+        userId: usersTrashUsage[index].userId,
+        open: usersTrashUsage[index].open.toISOString(),
+        trashLogs,
+      };
+    });
+
+    await Promise.all(
+      usersTrashLogsInTrial.map(({ userId, open, trashLogs }) => {
+        return Promise.all(
+          trashLogs.map(({ type, ok }) => {
+            const key = `user-trash:${userId}-${open}`;
+            const field = `${type}-${ok ? 'success' : 'failure'}`;
+            return this.client.hIncrBy(key, field, 1);
+          }),
+        );
+      }),
+    );
+  }
+
+  async getUserTrashMonthlySummary(
+    userId: number,
+    year: number,
+    month: number,
+  ): Promise<any> {
+    const userTrashUsagesInMonth = await this.client.lRange(
+      `user-trash:${userId}-${year}-${month}`,
+      0,
+      -1,
+    );
+    const userTrashLogsInMonth = (
+      await Promise.all(
+        userTrashUsagesInMonth.map((open) => {
+          return this.client.hGetAll(`user-trash:${userId}-${open}`);
+        }),
+      )
+    ).map((userTrashLogsInOneUsage, index) => {
+      const type = Object.keys(userTrashLogsInOneUsage)[0]?.split('-')[0];
+      const success = +(userTrashLogsInOneUsage[`${type}-success`] ?? 0);
+      const failure = +(userTrashLogsInOneUsage[`${type}-failure`] ?? 0);
+      return {
+        Date: Date.parse(userTrashUsagesInMonth[index]),
+        type,
+        success,
+        failure,
+      };
+    });
+    return userTrashLogsInMonth;
+  }
+
+  async addUserUsageTrialTrash(userId, open, close) {
+    const openString = format(open, 'yyyy-MM-dd HH:mm:ss.SSS', { locale: ko });
+    const closeString = format(close, 'yyyy-MM-dd HH:mm:ss.SSS', {
+      locale: ko,
+    });
+    const trashLogs = await this.databaseService.query<{
+      userId: string;
+      type: string;
+      at: Date;
+      ok: string;
+    }>(
+      `SELECT "userId", "type", "at", "ok" FROM ${database.tables.trash} 
+          WHERE "userId" = ${userId} 
+          AND "at" BETWEEN '${openString}' AND '${closeString}';`,
+    );
+    await Promise.all(
+      trashLogs.map(({ type, ok }) => {
+        const key = `user-trash:${userId}-${open}`;
         const field = `${type}-${ok ? 'success' : 'failure'}`;
         return this.client.hIncrBy(key, field, 1);
       }),
@@ -79,14 +201,8 @@ export class CacheService {
     });
   }
 
-  async getUserTrashSummary(
-    userId: number,
-    year: number,
-    month: number,
-  ): Promise<TrashSummary> {
-    const userTrashSummary = await this.client.hGetAll(
-      `user-trash:${userId}-${year}-${month}`,
-    );
+  async getUserTrashSummary(userId: number): Promise<TrashSummary> {
+    const userTrashSummary = await this.client.hGetAll(`user-trash:${userId}`);
     return {
       can: {
         success: +(userTrashSummary['can-success'] ?? 0),

--- a/src/cache/cache.service.ts
+++ b/src/cache/cache.service.ts
@@ -13,6 +13,10 @@ export class CacheService {
     private readonly databaseService: DatabaseService,
   ) {}
 
+  async flushAll() {
+    await this.client.flushAll();
+  }
+
   async migrateUsersPoint() {
     const usersPoint = await this.databaseService.query<{
       id: number;
@@ -69,7 +73,9 @@ export class CacheService {
           const openString = format(open, 'yyyy-MM-dd HH:mm:ss.SSS', {
             locale: ko,
           });
-          const key = `user-trash:${userId}-${open.getFullYear()}-${open.getMonth()}`;
+          const key = `user-trash:${userId}-${open.getFullYear()}-${
+            open.getMonth() + 1
+          }`;
           return this.client.rPush(key, openString);
         }),
       );
@@ -101,7 +107,9 @@ export class CacheService {
     ).map((trashLogs, index) => {
       return {
         userId: usersTrashUsage[index].userId,
-        open: usersTrashUsage[index].open.toISOString(),
+        open: format(usersTrashUsage[index].open, 'yyyy-MM-dd HH:mm:ss.SSS', {
+          locale: ko,
+        }),
         trashLogs,
       };
     });
@@ -129,6 +137,7 @@ export class CacheService {
       0,
       -1,
     );
+
     const userTrashLogsInMonth = (
       await Promise.all(
         userTrashUsagesInMonth.map((open) => {
@@ -140,7 +149,7 @@ export class CacheService {
       const success = +(userTrashLogsInOneUsage[`${type}-success`] ?? 0);
       const failure = +(userTrashLogsInOneUsage[`${type}-failure`] ?? 0);
       return {
-        Date: Date.parse(userTrashUsagesInMonth[index]),
+        Date: new Date(userTrashUsagesInMonth[index]),
         type,
         success,
         failure,

--- a/src/cache/cache.service.ts
+++ b/src/cache/cache.service.ts
@@ -157,6 +157,10 @@ export class CacheService {
     const [openString, closeString] = [open, close].map((date) => {
       return format(date, 'yyyy-MM-dd HH:mm:ss.SSS', { locale: ko });
     });
+    await this.client.rPush(
+      `user-trash:${userId}-${open.getFullYear()}-${open.getMonth() + 1}`,
+      openString,
+    );
     const trashLogs = await this.databaseService.query<{
       userId: string;
       type: string;
@@ -172,7 +176,7 @@ export class CacheService {
       failure = 0;
     await Promise.all(
       trashLogs.map(({ type: trashType, ok }) => {
-        const key = `user-trash:${userId}-${open}`;
+        const key = `user-trash:${userId}-${openString}`;
         const field = `${trashType}-${ok ? 'success' : 'failure'}`;
         type = trashType;
         ok ? success++ : failure++;

--- a/src/cache/cache.service.ts
+++ b/src/cache/cache.service.ts
@@ -171,6 +171,19 @@ export class CacheService {
         return this.client.hIncrBy(key, field, 1);
       }),
     );
+
+    const userTrashLogsInOneUsage = this.client.hKeys(
+      `user-trash:${userId}-${open}`,
+    );
+    const type = Object.keys(userTrashLogsInOneUsage)[0]?.split('-')[0];
+    const success = +(userTrashLogsInOneUsage[`${type}-success`] ?? 0);
+    const failure = +(userTrashLogsInOneUsage[`${type}-failure`] ?? 0);
+    return {
+      Date: open,
+      type,
+      success,
+      failure,
+    };
   }
 
   async addUserPoint(userId: number, point: number) {

--- a/src/common/environments.ts
+++ b/src/common/environments.ts
@@ -44,4 +44,9 @@ export const redis = {
   config: {
     url: `redis://${process.env.REDIS_HOST}:${process.env.REDIS_PORT}`,
   },
+  keys: {
+    userPoint: 'user-point',
+    userTrash: 'user-trash',
+    achievementNotification: 'achievement-notification',
+  },
 };

--- a/src/trash/dto/one-trial-trash-summary.dto.ts
+++ b/src/trash/dto/one-trial-trash-summary.dto.ts
@@ -1,0 +1,6 @@
+export class OneTrialTrashSummary {
+  date: Date;
+  type: string;
+  success: number;
+  failure: number;
+}

--- a/src/trash/dto/user-end-trash-response.dto.ts
+++ b/src/trash/dto/user-end-trash-response.dto.ts
@@ -1,0 +1,7 @@
+export class UserEndTrashRes {
+  totalPoint: number;
+  date: Date;
+  type: string;
+  success: number;
+  failure: number;
+}

--- a/src/trash/trash.controller.ts
+++ b/src/trash/trash.controller.ts
@@ -1,6 +1,7 @@
 import {
   Controller,
   Get,
+  NotFoundException,
   Param,
   ParseIntPipe,
   Post,
@@ -46,25 +47,23 @@ export class TrashController {
     return await this.trashService.endTrashcanUsage(usageId);
   }
 
-  @Get('summary')
-  @ApiOkResponse({
-    description: 'Get trash summary',
-    type: TrashSummary,
-  })
-  async summary(@AccessUser() user): Promise<TrashSummary> {
-    return await this.trashService.getUserTrashSummaryAll(user.id);
-  }
-
-  @Get('summary-monthly')
+  @Get('summary/:kind')
   @ApiOkResponse({
     description: 'Get trash summary',
     type: TrashSummary,
   })
   async summaryMonthly(
     @AccessUser() user,
+    @Param('kind') kind: string,
     @Query('year', ParseIntPipe) year: number,
     @Query('month', ParseIntPipe) month: number,
-  ): Promise<TrashSummary> {
-    return await this.trashService.getUserTrashSummaryInMonth(1, year, month);
+  ) {
+    if (kind === 'all') {
+      return await this.trashService.getUserTrashSummaryAll(1);
+    } else if (kind === 'detail') {
+      return await this.trashService.getUserTrashSummaryInMonth(1, year, month);
+    } else {
+      throw new NotFoundException();
+    }
   }
 }

--- a/src/trash/trash.controller.ts
+++ b/src/trash/trash.controller.ts
@@ -1,7 +1,7 @@
 import {
   Controller,
+  DefaultValuePipe,
   Get,
-  NotFoundException,
   Param,
   ParseIntPipe,
   Post,
@@ -14,6 +14,8 @@ import { JwtAuthGuard } from 'src/auth/guard/jwt-auth.guard';
 import { Trashcan } from 'src/trashcans/entities/trashcan.entity';
 import { TrashcanByIdPipe } from 'src/trashcans/pipe/trashcan-by-id.pipe';
 import { TrashSummary } from './dto/trash-summary.dto';
+import { UserEndTrashRes } from './dto/user-end-trash-response.dto';
+import { OneTrialTrashSummary } from './dto/one-trial-trash-summary.dto';
 import { TrashService } from './trash.service';
 
 @ApiTags('Trash')
@@ -43,31 +45,29 @@ export class TrashController {
   async end(
     @AccessUser() user,
     @Query('usageId', ParseIntPipe) usageId: number,
-  ) {
+  ): Promise<UserEndTrashRes> {
     return await this.trashService.endTrashcanUsage(usageId);
   }
 
-  @Get('summary/:kind')
+  @Get('summary/all')
   @ApiOkResponse({
     description: 'Get trash summary',
     type: TrashSummary,
   })
-  async summaryMonthly(
+  async allSummary(@AccessUser() user): Promise<TrashSummary> {
+    return await this.trashService.getUserTrashSummaryAll(user.id);
+  }
+
+  @Get('summary/detail')
+  async detailSummary(
     @AccessUser() user,
-    @Param('kind') kind: string,
-    @Query('year', ParseIntPipe) year: number,
-    @Query('month', ParseIntPipe) month: number,
-  ) {
-    if (kind === 'all') {
-      return await this.trashService.getUserTrashSummaryAll(user.id);
-    } else if (kind === 'detail') {
-      return await this.trashService.getUserTrashSummaryInMonth(
-        user.id,
-        year,
-        month,
-      );
-    } else {
-      throw new NotFoundException();
-    }
+    @Query('year', ParseIntPipe, new DefaultValuePipe(0)) year: number,
+    @Query('month', ParseIntPipe, new DefaultValuePipe(0)) month: number,
+  ): Promise<OneTrialTrashSummary[]> {
+    return await this.trashService.getUserTrashSummaryDetail(
+      user.id,
+      year,
+      month,
+    );
   }
 }

--- a/src/trash/trash.controller.ts
+++ b/src/trash/trash.controller.ts
@@ -29,12 +29,12 @@ export class TrashController {
   })
   async begin(
     @AccessUser() user,
-    //@Param('id', TrashcanByIdPipe) trashcan: Trashcan,
+    @Param('id', TrashcanByIdPipe) trashcan: Trashcan,
   ): Promise<number> {
-    return await this.trashService.beginTrashcanUsage(user.id, 1);
+    return await this.trashService.beginTrashcanUsage(user.id, trashcan.id);
   }
 
-  @Post('end/:id')
+  @Post('end')
   @ApiOkResponse({
     description: 'End trashcan usage',
     type: Boolean,

--- a/src/trash/trash.controller.ts
+++ b/src/trash/trash.controller.ts
@@ -59,9 +59,13 @@ export class TrashController {
     @Query('month', ParseIntPipe) month: number,
   ) {
     if (kind === 'all') {
-      return await this.trashService.getUserTrashSummaryAll(1);
+      return await this.trashService.getUserTrashSummaryAll(user.id);
     } else if (kind === 'detail') {
-      return await this.trashService.getUserTrashSummaryInMonth(1, year, month);
+      return await this.trashService.getUserTrashSummaryInMonth(
+        user.id,
+        year,
+        month,
+      );
     } else {
       throw new NotFoundException();
     }

--- a/src/trash/trash.controller.ts
+++ b/src/trash/trash.controller.ts
@@ -29,7 +29,7 @@ export class TrashController {
   async begin(
     @AccessUser() user,
     @Param('id', TrashcanByIdPipe) trashcan: Trashcan,
-  ): Promise<boolean> {
+  ): Promise<number> {
     return await this.trashService.beginTrashcanUsage(user.id, trashcan.id);
   }
 
@@ -41,8 +41,9 @@ export class TrashController {
   async end(
     @AccessUser() user,
     @Param('id', TrashcanByIdPipe) trashcan: Trashcan,
-  ): Promise<boolean> {
-    return await this.trashService.endTrashcanUsage(user.id, trashcan.id);
+    @Query('usageId', ParseIntPipe) usageId: number,
+  ) {
+    return await this.trashService.endTrashcanUsage(usageId);
   }
 
   @Get('summary')
@@ -50,11 +51,20 @@ export class TrashController {
     description: 'Get trash summary',
     type: TrashSummary,
   })
-  async summary(
+  async summary(@AccessUser() user): Promise<TrashSummary> {
+    return await this.trashService.getUserTrashSummaryAll(user.id);
+  }
+
+  @Get('summary-monthly')
+  @ApiOkResponse({
+    description: 'Get trash summary',
+    type: TrashSummary,
+  })
+  async summaryMonthly(
     @AccessUser() user,
-    @Query('year', ParseIntPipe) year,
-    @Query('month', ParseIntPipe) month,
+    @Query('year', ParseIntPipe) year: number,
+    @Query('month', ParseIntPipe) month: number,
   ): Promise<TrashSummary> {
-    return await this.trashService.getUserTrashSummary(user.id, year, month);
+    return await this.trashService.getUserTrashSummaryInMonth(1, year, month);
   }
 }

--- a/src/trash/trash.controller.ts
+++ b/src/trash/trash.controller.ts
@@ -7,7 +7,7 @@ import {
   Query,
   UseGuards,
 } from '@nestjs/common';
-import { ApiOkResponse, ApiTags } from '@nestjs/swagger';
+import { ApiBearerAuth, ApiOkResponse, ApiTags } from '@nestjs/swagger';
 import { AccessUser } from 'src/auth/decorator/access-user.decorator';
 import { JwtAuthGuard } from 'src/auth/guard/jwt-auth.guard';
 import { Trashcan } from 'src/trashcans/entities/trashcan.entity';
@@ -18,6 +18,7 @@ import { TrashService } from './trash.service';
 @ApiTags('Trash')
 @Controller('trash')
 @UseGuards(JwtAuthGuard)
+@ApiBearerAuth()
 export class TrashController {
   constructor(private readonly trashService: TrashService) {}
 
@@ -28,9 +29,9 @@ export class TrashController {
   })
   async begin(
     @AccessUser() user,
-    @Param('id', TrashcanByIdPipe) trashcan: Trashcan,
+    //@Param('id', TrashcanByIdPipe) trashcan: Trashcan,
   ): Promise<number> {
-    return await this.trashService.beginTrashcanUsage(user.id, trashcan.id);
+    return await this.trashService.beginTrashcanUsage(user.id, 1);
   }
 
   @Post('end/:id')
@@ -40,7 +41,6 @@ export class TrashController {
   })
   async end(
     @AccessUser() user,
-    @Param('id', TrashcanByIdPipe) trashcan: Trashcan,
     @Query('usageId', ParseIntPipe) usageId: number,
   ) {
     return await this.trashService.endTrashcanUsage(usageId);

--- a/src/trash/trash.module.ts
+++ b/src/trash/trash.module.ts
@@ -1,3 +1,4 @@
+import { UsersModule } from 'src/users/users.module';
 import { Module } from '@nestjs/common';
 import { TrashController } from './trash.controller';
 import { TrashRepository } from './trash.repository';
@@ -7,7 +8,7 @@ import { TrashcansModule } from 'src/trashcans/trashcans.module';
 import { DatabaseModule } from 'src/database/database.module';
 
 @Module({
-  imports: [CacheModule, TrashcansModule, DatabaseModule],
+  imports: [CacheModule, TrashcansModule, DatabaseModule, UsersModule],
   controllers: [TrashController],
   providers: [TrashService, TrashRepository],
 })

--- a/src/trash/trash.repository.ts
+++ b/src/trash/trash.repository.ts
@@ -6,12 +6,11 @@ import { Injectable } from '@nestjs/common';
 export class TrashRepository {
   constructor(private databaseService: DatabaseService) {}
 
-  async logTrashcanUsage(
+  async createTrashcanUsage(
     userId: number,
     trashcanId: number,
-    open: boolean,
-  ): Promise<boolean> {
-    const queryResult = await this.databaseService.query<{ id: string }>(`
+  ): Promise<number> {
+    const queryResult = await this.databaseService.query<{ id: number }>(`
       INSERT INTO
         ${database.tables.trashcanUsage}
         (
@@ -23,12 +22,31 @@ export class TrashRepository {
         (
           ${userId},
           ${trashcanId},
-          ${open}
+          now()
         )
       RETURNING
         id
     ;`);
 
-    return queryResult.length === 1;
+    return queryResult.length === 1 ? +queryResult[0].id : null;
+  }
+
+  async updateTrashcanUsage(usageId: number) {
+    const queryResult = await this.databaseService.query<{
+      userId: number;
+      open: Date;
+      close: Date;
+    }>(`
+      UPDATE
+        ${database.tables.trashcanUsage}
+      SET
+        "close" = now()
+      WHERE
+        id = ${usageId}
+      RETURNING
+        "userId","open","close"
+    ;`);
+
+    return queryResult.length === 1 ? queryResult[0] : null;
   }
 }

--- a/src/trash/trash.service.ts
+++ b/src/trash/trash.service.ts
@@ -1,3 +1,4 @@
+import { UsersRepository } from 'src/users/users.repository';
 import { Injectable } from '@nestjs/common';
 import { CacheService } from 'src/cache/cache.service';
 import { TrashSummary } from './dto/trash-summary.dto';
@@ -8,6 +9,7 @@ export class TrashService {
   constructor(
     private readonly trashRepository: TrashRepository,
     private readonly cacheService: CacheService,
+    private readonly usersRepository: UsersRepository,
   ) {}
 
   async beginTrashcanUsage(
@@ -20,7 +22,16 @@ export class TrashService {
   async endTrashcanUsage(usageId: number) {
     const { userId, open, close } =
       await this.trashRepository.updateTrashcanUsage(usageId);
-    return this.cacheService.addUserUsageTrialTrash(userId, open, close);
+    const UserUsageTrialTrash = this.cacheService.addUserUsageTrialTrash(
+      userId,
+      open,
+      close,
+    );
+    const totalPoint = await (
+      await this.usersRepository.findByUserId(userId)
+    ).point;
+
+    return { ...UserUsageTrialTrash, totalPoint };
   }
 
   async getUserTrashSummaryAll(userId: number): Promise<TrashSummary> {

--- a/src/trash/trash.service.ts
+++ b/src/trash/trash.service.ts
@@ -13,27 +13,29 @@ export class TrashService {
   async beginTrashcanUsage(
     userId: number,
     trashcanId: number,
-  ): Promise<boolean> {
-    return await this.trashRepository.logTrashcanUsage(
-      userId,
-      trashcanId,
-      true,
-    );
+  ): Promise<number> {
+    return await this.trashRepository.createTrashcanUsage(userId, trashcanId);
   }
 
-  async endTrashcanUsage(userId: number, trashcanId: number): Promise<boolean> {
-    return await this.trashRepository.logTrashcanUsage(
-      userId,
-      trashcanId,
-      false,
-    );
+  async endTrashcanUsage(usageId: number) {
+    const { userId, open, close } =
+      await this.trashRepository.updateTrashcanUsage(usageId);
+    await this.cacheService.addUserUsageTrialTrash(userId, open, close);
   }
 
-  async getUserTrashSummary(
+  async getUserTrashSummaryAll(userId: number): Promise<TrashSummary> {
+    return await this.cacheService.getUserTrashSummary(userId);
+  }
+
+  async getUserTrashSummaryInMonth(
     userId: number,
     year: number,
     month: number,
   ): Promise<TrashSummary> {
-    return await this.cacheService.getUserTrashSummary(userId, year, month);
+    return await this.cacheService.getUserTrashMonthlySummary(
+      userId,
+      year,
+      month,
+    );
   }
 }

--- a/src/trash/trash.service.ts
+++ b/src/trash/trash.service.ts
@@ -31,7 +31,7 @@ export class TrashService {
     userId: number,
     year: number,
     month: number,
-  ): Promise<TrashSummary> {
+  ) {
     return await this.cacheService.getUserTrashMonthlySummary(
       userId,
       year,

--- a/src/trash/trash.service.ts
+++ b/src/trash/trash.service.ts
@@ -1,8 +1,10 @@
+import { OneTrialTrashSummary } from './dto/one-trial-trash-summary.dto';
 import { UsersRepository } from 'src/users/users.repository';
 import { Injectable } from '@nestjs/common';
 import { CacheService } from 'src/cache/cache.service';
 import { TrashSummary } from './dto/trash-summary.dto';
 import { TrashRepository } from './trash.repository';
+import { UserEndTrashRes } from './dto/user-end-trash-response.dto';
 
 @Injectable()
 export class TrashService {
@@ -19,7 +21,7 @@ export class TrashService {
     return await this.trashRepository.createTrashcanUsage(userId, trashcanId);
   }
 
-  async endTrashcanUsage(usageId: number) {
+  async endTrashcanUsage(usageId: number): Promise<UserEndTrashRes> {
     const { userId, open, close } =
       await this.trashRepository.updateTrashcanUsage(usageId);
     const UserUsageTrialTrash = await this.cacheService.addUserUsageTrialTrash(
@@ -27,22 +29,26 @@ export class TrashService {
       open,
       close,
     );
-    const totalPoint = await (
-      await this.usersRepository.findByUserId(userId)
-    ).point;
+    const totalPoint = (await this.usersRepository.findByUserId(userId)).point;
+    await this.cacheService.updateUserTrashAllSummary(
+      userId,
+      UserUsageTrialTrash.type,
+      UserUsageTrialTrash.success,
+      UserUsageTrialTrash.failure,
+    );
 
     return { ...UserUsageTrialTrash, totalPoint };
   }
 
   async getUserTrashSummaryAll(userId: number): Promise<TrashSummary> {
-    return await this.cacheService.getUserTrashSummary(userId);
+    return await this.cacheService.getUserTrashAllSummary(userId);
   }
 
-  async getUserTrashSummaryInMonth(
+  async getUserTrashSummaryDetail(
     userId: number,
     year: number,
     month: number,
-  ) {
+  ): Promise<OneTrialTrashSummary[]> {
     return await this.cacheService.getUserTrashMonthlySummary(
       userId,
       year,

--- a/src/trash/trash.service.ts
+++ b/src/trash/trash.service.ts
@@ -20,7 +20,7 @@ export class TrashService {
   async endTrashcanUsage(usageId: number) {
     const { userId, open, close } =
       await this.trashRepository.updateTrashcanUsage(usageId);
-    await this.cacheService.addUserUsageTrialTrash(userId, open, close);
+    return this.cacheService.addUserUsageTrialTrash(userId, open, close);
   }
 
   async getUserTrashSummaryAll(userId: number): Promise<TrashSummary> {

--- a/src/trash/trash.service.ts
+++ b/src/trash/trash.service.ts
@@ -22,7 +22,7 @@ export class TrashService {
   async endTrashcanUsage(usageId: number) {
     const { userId, open, close } =
       await this.trashRepository.updateTrashcanUsage(usageId);
-    const UserUsageTrialTrash = this.cacheService.addUserUsageTrialTrash(
+    const UserUsageTrialTrash = await this.cacheService.addUserUsageTrialTrash(
       userId,
       open,
       close,

--- a/src/trashcans/trashcans.repository.ts
+++ b/src/trashcans/trashcans.repository.ts
@@ -3,12 +3,12 @@ import { DatabaseService } from 'src/database/database.service';
 import { CreateTrashcanDto } from './dto/create-trashcan.dto';
 import { UpdateTrashcanDto } from './dto/update-trashcan.dto';
 import { Trashcan } from './entities/trashcan.entity';
+import { database } from 'src/common/environments';
 
 @Injectable()
 export class TrashcansRepository {
-  private readonly schema = process.env.DATABASE_APPLICATION_SCHEMA;
-  private readonly trashcanTable = `${this.schema}.trashcan`;
-  private readonly trashcanCodeTable = `"${this.schema}"."trashcanCode"`;
+  private readonly trashcanTable = database.tables.trashcan;
+  private readonly trashcanCodeTable = database.tables.trashcanCode;
 
   constructor(private readonly databaseService: DatabaseService) {}
 

--- a/src/users/dto/user-response.dto.ts
+++ b/src/users/dto/user-response.dto.ts
@@ -6,7 +6,7 @@ import {
   IsString,
   ValidateNested,
 } from 'class-validator';
-import { Achievement } from '../entities/achievement.entity';
+import { UserAchievement } from 'src/achievements/entity/user-achievement.entity';
 import { Grade } from '../types/grade.enum';
 
 export class UserResponseDto {
@@ -33,9 +33,9 @@ export class UserResponseDto {
   @IsNotEmpty()
   @ValidateNested({ each: true })
   @Type(() => {
-    return Achievement;
+    return UserAchievement;
   })
-  achievement: Achievement[];
+  achievement: UserAchievement[];
 
   @IsNotEmpty()
   @IsNumber()

--- a/src/users/users.module.ts
+++ b/src/users/users.module.ts
@@ -1,3 +1,4 @@
+import { AchievementsModule } from 'src/achievements/achievements.module';
 import { CacheModule } from './../cache/cache.module';
 import { forwardRef, Module } from '@nestjs/common';
 import { UsersService } from './users.service';
@@ -13,6 +14,7 @@ import { UsersRepository } from './users.repository';
     }),
     DatabaseModule,
     CacheModule,
+    AchievementsModule,
   ],
   controllers: [UsersController],
   providers: [UsersService, UsersRepository],

--- a/src/users/users.repository.ts
+++ b/src/users/users.repository.ts
@@ -2,7 +2,7 @@ import { User } from 'src/users/entities/user.entity';
 import { Injectable } from '@nestjs/common';
 import { DatabaseService } from 'src/database/database.service';
 import { database } from 'src/common/environments';
-import { Achievement } from './entities/achievement.entity';
+import { UserAchievement } from 'src/achievements/entity/user-achievement.entity';
 
 @Injectable()
 export class UsersRepository {
@@ -31,16 +31,6 @@ export class UsersRepository {
       `SELECT * FROM ${database.tables.user} WHERE id=${userId};`,
     );
     return result.length === 1 ? result[0] : null;
-  }
-
-  async findAcheiveByUserId(userId: number): Promise<Achievement[]> {
-    const result = await this.databaseService.query<Achievement>(
-      `SELECT ${database.tables.achievement}.*,${database.tables.achiever}."achievedAt"
-        FROM ${database.tables.achievement} LEFT JOIN ${database.tables.achiever}
-            ON ${database.tables.achievement}.id = ${database.tables.achiever}."achievementId"
-            AND ${database.tables.achiever}."userId" = ${userId};`,
-    );
-    return result;
   }
 
   async update(name: string, thumbnail: string, id: number): Promise<User> {

--- a/src/users/users.service.ts
+++ b/src/users/users.service.ts
@@ -1,18 +1,20 @@
+import { AchievementsRepository } from 'src/achievements/achievements.repository';
 import { CacheService } from './../cache/cache.service';
 import { User } from 'src/users/entities/user.entity';
 import { UsersRepository } from './users.repository';
 import { Injectable } from '@nestjs/common';
-import { Achievement } from './entities/achievement.entity';
 import { UserResponseDto } from './dto/user-response.dto';
 import { UserUpdateDto } from './dto/user-update.dto';
 import { UserRankResponseDto } from './dto/user-rank-response.dto';
 import { Grade } from './types/grade.enum';
+import { UserAchievement } from 'src/achievements/entity/user-achievement.entity';
 
 @Injectable()
 export class UsersService {
   constructor(
     private usersRepository: UsersRepository,
     private cacheService: CacheService,
+    private achievementsRepository: AchievementsRepository,
   ) {}
 
   async create(kakaoId: bigint, thumbnail: string, name: string) {
@@ -36,8 +38,8 @@ export class UsersService {
         ? Grade.TheVolunteer
         : Grade.TheBeginner;
 
-    const achievement: Achievement[] =
-      await this.usersRepository.findAcheiveByUserId(id);
+    const achievement: UserAchievement[] =
+      await this.achievementsRepository.getUserAchievements(id);
 
     const rank = await this.cacheService.getUserRank(id);
 

